protocol-tokio/src/protocol/mod.rs

```diff
@@ -65,16 +66,26 @@ impl ConnectionState {
 ///
 /// Once established call `split` to get the inbound stream
 /// and the outbound sink and starts processing queries
-pub struct Connection<T> {
+pub struct Connection<T, Header, BlockId, Block, TransactionId> {
     connection: nt::Connection<T>,
     state: Arc<Mutex<ConnectionState>>,
+    header: PhantomData<Header>,
+    block_id: PhantomData<BlockId>,
+    block: PhantomData<Block>,
+    transaction_id: PhantomData<TransactionId>,
```

Comment:
This is annoying. Maybe we should add a trait to network-core grouping all these types as associated, to have just one parameter here and in the protocol message types above.